velox/connectors/hive/iceberg/PositionalDeleteFileReader.cpp

```diff
@@ -0,0 +1,264 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/iceberg/PositionalDeleteFileReader.h"
+
+#include "velox/connectors/hive/HiveConnectorSplit.h"
+#include "velox/connectors/hive/HiveConnectorUtil.h"
+#include "velox/connectors/hive/TableHandle.h"
+#include "velox/connectors/hive/iceberg/IcebergDeleteFile.h"
+#include "velox/connectors/hive/iceberg/IcebergMetadataColumns.h"
+#include "velox/dwio/common/ReaderFactory.h"
+
+namespace facebook::velox::connector::hive::iceberg {
+
+PositionalDeleteFileReader::PositionalDeleteFileReader(
+    const IcebergDeleteFile& deleteFile,
+    const std::string& baseFilePath,
+    FileHandleFactory* fileHandleFactory,
+    const ConnectorQueryCtx* connectorQueryCtx,
+    folly::Executor* executor,
+    const std::shared_ptr<HiveConfig> hiveConfig,
+    std::shared_ptr<io::IoStatistics> ioStats,
+    dwio::common::RuntimeStatistics& runtimeStats,
+    uint64_t splitOffset,
+    const std::string& connectorId)
+    : deleteFile_(deleteFile),
+      baseFilePath_(baseFilePath),
+      fileHandleFactory_(fileHandleFactory),
+      executor_(executor),
+      connectorQueryCtx_(connectorQueryCtx),
+      ioStats_(ioStats),
+      pool_(connectorQueryCtx->memoryPool()),
+      filePathColumn_(ICEBERG_DELETE_FILE_PATH_COLUMN()),
+      posColumn_(ICEBERG_DELETE_FILE_POSITIONS_COLUMN()),
+      splitOffset_(splitOffset),
+      deleteSplit_(nullptr),
+      deleteRowReader_(nullptr),
+      deletePositionsOutput_(nullptr),
+      deletePositionsOffset_(0),
+      endOfFile_(false) {
+  VELOX_CHECK_EQ(deleteFile_.content, FileContent::kPositionalDeletes);
+
+  if (deleteFile_.recordCount == 0) {
+    return;
+  }
+
+  // TODO: check if the lowerbounds and upperbounds in deleteFile overlap with
+  //  this batch. If not, no need to proceed.
+
+  std::vector<std::string> deleteColumnNames(
+      {filePathColumn_->name, posColumn_->name});
+  std::vector<std::shared_ptr<const Type>> deleteColumnTypes(
+      {filePathColumn_->type, posColumn_->type});
+  RowTypePtr deleteFileSchema =
+      ROW(std::move(deleteColumnNames), std::move(deleteColumnTypes));
+  auto outputRowType = ROW({posColumn_->name}, {posColumn_->type});
+
+  common::Subfield posSubfield(posColumn_->name);
+  folly::F14FastMap<std::string, std::vector<const common::Subfield*>>
+      outputSubfieldsMap;
+  outputSubfieldsMap[posColumn_->name].push_back(&posSubfield);
+
+  size_t lastPathDelimiterPos = deleteFile_.filePath.find_last_of('/');
+  std::string deleteFileName = deleteFile_.filePath.substr(
+      lastPathDelimiterPos, deleteFile_.filePath.size() - lastPathDelimiterPos);
+  auto deleteTableHandle = std::make_shared<HiveTableHandle>(
+      connectorId, deleteFileName, false, createFilters(), nullptr);
+
+  auto scanSpec = makeScanSpec(
```

Comment:
You don't need `makeScanSpec`, just make `createFilters()` return the scan spec with filters set as I mentioned here: https://github.com/facebookincubator/velox/pull/7362#discussion_r1412317067

Reply:
@Yuhta I tried what you suggested, but it throws exception because of the channel out of bound for the _pos column. Note that we only project out _pos column, but when we add it through addAllChildFields(), it assigns the channel 1 for _pos column. Adding more lines to fix this may not be as neat as just calling makeScanSpec so I converted it back to makeScanSpec().